File: packages/go/msgpack.go
package flatwire

// Streaming MessagePack (binary) format for flatwire, mirroring the Python
// reference. flatwire's binary wire is a stream of concatenated MessagePack
// values (not a length-prefixed array), so encoding needs no upfront count and
// decoding reads one value at a time. Wire-compatible with standard MessagePack
// for the JSON data model (null/bool/int/float/str/array/map); no ext types.
//
// Values are Go's generic any: objects -> map[string]any, arrays -> []any,
// scalars -> int64/uint64/float64/bool/string/nil.

import (
	"bufio"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"math"
)

// --- encoding --------------------------------------------------------------

func writeMsgValue(w io.Writer, v any) error {
	switch x := v.(type) {
	case nil:
		_, err := w.Write([]byte{0xc0})
		return err
	case bool:
		if x {
			_, err := w.Write([]byte{0xc3})
			return err
		}
		_, err := w.Write([]byte{0xc2})
		return err
	case string:
		return writeMsgStr(w, x)
	case float32:
		return writeMsgF64(w, float64(x))
	case float64:
		return writeMsgF64(w, x)
	case int:
		return writeMsgInt(w, int64(x))
	case int8:
		return writeMsgInt(w, int64(x))
	case int16:
		return writeMsgInt(w, int64(x))
	case int32:
		return writeMsgInt(w, int64(x))
	case int64:
		return writeMsgInt(w, x)
	case uint:
		return writeMsgUint(w, uint64(x))
	case uint8:
		return writeMsgUint(w, uint64(x))
	case uint16:
		return writeMsgUint(w, uint64(x))
	case uint32:
		return writeMsgUint(w, uint64(x))
	case uint64:
		return writeMsgUint(w, x)
	case []any:
		if err := writeMsgArrayHeader(w, len(x)); err != nil {
			return err
		}
		for _, e := range x {
			if err := writeMsgValue(w, e); err != nil {
				return err
			}
		}
		return nil
	case map[string]any:
		if err := writeMsgMapHeader(w, len(x)); err != nil {
			return err
		}
		for k, val := range x {
			if err := writeMsgStr(w, k); err != nil {
				return err
			}
			if err := writeMsgValue(w, val); err != nil {
				return err
			}
		}
		return nil
	default:
		return fmt.Errorf("flatwire msgpack: unsupported type %T", v)
	}
}

func writeMsgF64(w io.Writer, f float64) error {
	var b [9]byte
	b[0] = 0xcb
	binary.BigEndian.PutUint64(b[1:], math.Float64bits(f))
	_, err := w.Write(b[:])
	return err
}

func writeMsgInt(w io.Writer, v int64) error {
	switch {
	case v >= 0 && v <= 0x7f:
		_, err := w.Write([]byte{byte(v)})
		return err
	case v < 0 && v >= -32:
		_, err := w.Write([]byte{byte(v)})
		return err
	case v >= -0x80 && v <= 0x7f:
		_, err := w.Write([]byte{0xd0, byte(v)})
		return err
	case v >= -0x8000 && v <= 0x7fff:
		var b [3]byte
		b[0] = 0xd1
		binary.BigEndian.PutUint16(b[1:], uint16(int16(v)))
		_, err := w.Write(b[:])
		return err
	case v >= -0x80000000 && v <= 0x7fffffff:
		var b [5]byte
		b[0] = 0xd2
		binary.BigEndian.PutUint32(b[1:], uint32(int32(v)))
		_, err := w.Write(b[:])
		return err
	default:
		var b [9]byte
		b[0] = 0xd3
		binary.BigEndian.PutUint64(b[1:], uint64(v))
		_, err := w.Write(b[:])
		return err
	}
}

func writeMsgUint(w io.Writer, v uint64) error {
	switch {
	case v <= 0x7f:
		_, err := w.Write([]byte{byte(v)})
		return err
	case v <= 0xff:
		_, err := w.Write([]byte{0xcc, byte(v)})
		return err
	case v <= 0xffff:
		var b [3]byte
		b[0] = 0xcd
		binary.BigEndian.PutUint16(b[1:], uint16(v))
		_, err := w.Write(b[:])
		return err
	case v <= 0xffffffff:
		var b [5]byte
		b[0] = 0xce
		binary.BigEndian.PutUint32(b[1:], uint32(v))
		_, err := w.Write(b[:])
		return err
	default:
		var b [9]byte
		b[0] = 0xcf
		binary.BigEndian.PutUint64(b[1:], v)
		_, err := w.Write(b[:])
		return err
	}
}

func writeMsgStr(w io.Writer, s string) error {
	n := len(s)
	switch {
	case n <= 31:
		if _, err := w.Write([]byte{0xa0 | byte(n)}); err != nil {
			return err
		}
	case n <= 0xff:
		if _, err := w.Write([]byte{0xd9, byte(n)}); err != nil {
			return err
		}
	case n <= 0xffff:
		var b [3]byte
		b[0] = 0xda
		binary.BigEndian.PutUint16(b[1:], uint16(n))
		if _, err := w.Write(b[:]); err != nil {
			return err
		}
	default:
		var b [5]byte
		b[0] = 0xdb
		binary.BigEndian.PutUint32(b[1:], uint32(n))
		if _, err := w.Write(b[:]); err != nil {
			return err
		}
	}
	_, err := io.WriteString(w, s)
	return err
}

func writeMsgArrayHeader(w io.Writer, n int) error {
	switch {
	case n <= 15:
		_, err := w.Write([]byte{0x90 | byte(n)})
		return err
	case n <= 0xffff:
		var b [3]byte
		b[0] = 0xdc
		binary.BigEndian.PutUint16(b[1:], uint16(n))
		_, err := w.Write(b[:])
		return err
	default:
		var b [5]byte
		b[0] = 0xdd
		binary.BigEndian.PutUint32(b[1:], uint32(n))
		_, err := w.Write(b[:])
		return err
	}
}

func writeMsgMapHeader(w io.Writer, n int) error {
	switch {
	case n <= 15:
		_, err := w.Write([]byte{0x80 | byte(n)})
		return err
	case n <= 0xffff:
		var b [3]byte
		b[0] = 0xde
		binary.BigEndian.PutUint16(b[1:], uint16(n))
		_, err := w.Write(b[:])
		return err
	default:
		var b [5]byte
		b[0] = 0xdf
		binary.BigEndian.PutUint32(b[1:], uint32(n))
		_, err := w.Write(b[:])
		return err
	}
}

// EncodeArrayMsgPack streams a collection as concatenated MessagePack values, one
// per element. Peak memory is bounded by the largest single element. Returns the
// element count.
func EncodeArrayMsgPack(items []any, w io.Writer) (int, error) {
	count := 0
	for _, v := range items {
		if err := writeMsgValue(w, v); err != nil {
			return count, err
		}
		count++
	}
	return count, nil
}

// --- decoding --------------------------------------------------------------

// DecodeArrayMsgPack lazily reads concatenated MessagePack values, calling yield
// with each decoded element in turn. Backed by a buffered reader, so the whole
// stream is never held in memory at once.
func DecodeArrayMsgPack(r io.Reader, yield func(any) error) error {
	br := bufio.NewReaderSize(r, 65536)
	for {
		_, err := br.Peek(1)
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		v, err := readMsgValue(br)
		if err != nil {
			return err
		}
		if err := yield(v); err != nil {
			return err
		}
	}
}

func readMsgValue(br *bufio.Reader) (any, error) {
	c, err := br.ReadByte()
	if err != nil {
		return nil, err
	}
	switch {
	case c <= 0x7f:
		return int64(c), nil
	case c >= 0xe0:
		return int64(int8(c)), nil
	case c >= 0x80 && c <= 0x8f:
		return readMsgMap(br, int(c&0x0f))
	case c >= 0x90 && c <= 0x9f:
		return readMsgArray(br, int(c&0x0f))
	case c >= 0xa0 && c <= 0xbf:
		return readMsgStr(br, int(c&0x1f))
	}
	switch c {
	case 0xc0:
		return nil, nil
	case 0xc2:
		return false, nil
	case 0xc3:
		return true, nil
	case 0xca:
		b, err := readN(br, 4)
		if err != nil {
			return nil, err
		}
		return float64(math.Float32frombits(binary.BigEndian.Uint32(b))), nil
	case 0xcb:
		b, err := readN(br, 8)
		if err != nil {
			return nil, err
		}
		return math.Float64frombits(binary.BigEndian.Uint64(b)), nil
	case 0xcc:
		b, err := br.ReadByte()
		return int64(b), err
	case 0xcd:
		b, err := readN(br, 2)
		if err != nil {
			return nil, err
		}
		return int64(binary.BigEndian.Uint16(b)), nil
	case 0xce:
		b, err := readN(br, 4)
		if err != nil {
			return nil, err
		}
		return int64(binary.BigEndian.Uint32(b)), nil
	case 0xcf:
		b, err := readN(br, 8)
		if err != nil {
			return nil, err
		}
		return binary.BigEndian.Uint64(b), nil
	case 0xd0:
		b, err := br.ReadByte()
		return int64(int8(b)), err
	case 0xd1:
		b, err := readN(br, 2)
		if err != nil {
			return nil, err
		}
		return int64(int16(binary.BigEndian.Uint16(b))), nil
	case 0xd2:
		b, err := readN(br, 4)
		if err != nil {
			return nil, err
		}
		return int64(int32(binary.BigEndian.Uint32(b))), nil
	case 0xd3:
		b, err := readN(br, 8)
		if err != nil {
			return nil, err
		}
		return int64(binary.BigEndian.Uint64(b)), nil
	case 0xd9:
		n, err := br.ReadByte()
		if err != nil {
			return nil, err
		}
		return readMsgStr(br, int(n))
	case 0xda:
		b, err := readN(br, 2)
		if err != nil {
			return nil, err
		}
		return readMsgStr(br, int(binary.BigEndian.Uint16(b)))
	case 0xdb:
		b, err := readN(br, 4)
		if err != nil {
			return nil, err
		}
		return readMsgStr(br, int(binary.BigEndian.Uint32(b)))
	case 0xdc:
		b, err := readN(br, 2)
		if err != nil {
			return nil, err
		}
		return readMsgArray(br, int(binary.BigEndian.Uint16(b)))
	case 0xdd:
		b, err := readN(br, 4)
		if err != nil {
			return nil, err
		}
		return readMsgArray(br, int(binary.BigEndian.Uint32(b)))
	case 0xde:
		b, err := readN(br, 2)
		if err != nil {
			return nil, err
		}
		return readMsgMap(br, int(binary.BigEndian.Uint16(b)))
	case 0xdf:
		b, err := readN(br, 4)
		if err != nil {
			return nil, err
		}
		return readMsgMap(br, int(binary.BigEndian.Uint32(b)))
	case 0xc4, 0xc5, 0xc6:
		return nil, errors.New("flatwire msgpack: binary (bin) type is not part of the JSON value model")
	}
	return nil, fmt.Errorf("flatwire msgpack: unknown prefix 0x%02x", c)
}

func readN(br *bufio.Reader, n int) ([]byte, error) {
	b := make([]byte, n)
	_, err := io.ReadFull(br, b)
	return b, err
}

func readMsgStr(br *bufio.Reader, n int) (any, error) {
	b, err := readN(br, n)
	if err != nil {
		return nil, err
	}
	return string(b), nil
}

func readMsgArray(br *bufio.Reader, n int) (any, error) {
	arr := make([]any, n)
	for i := 0; i < n; i++ {
		v, err := readMsgValue(br)
		if err != nil {
			return nil, err
		}
		arr[i] = v
	}
	return arr, nil
}

func readMsgMap(br *bufio.Reader, n int) (any, error) {
	m := make(map[string]any, n)
	for i := 0; i < n; i++ {
		k, err := readMsgValue(br)
		if err != nil {
			return nil, err
		}
		v, err := readMsgValue(br)
		if err != nil {
			return nil, err
		}
		key, ok := k.(string)
		if !ok {
			key = fmt.Sprintf("%v", k)
		}
		m[key] = v
	}
	return m, nil
}

File: packages/rust/src/msgpack.rs
//! Streaming MessagePack (binary) format for flatwire, mirroring the Python
//! reference. flatwire's binary wire is a stream of concatenated MessagePack
//! values (not a length-prefixed array), so encoding needs no upfront count and
//! decoding reads one value at a time. Wire-compatible with standard MessagePack
//! for the JSON data model (null/bool/int/float/str/bin/array/map); no ext types.
//!
//! Values are `serde_json::Value`, keeping the binary path type-uniform with the
//! JSON and XML paths.

use std::io::{self, Read, Write};

use serde_json::{Map, Value};

// --- encoding --------------------------------------------------------------

fn write_value<W: Write>(w: &mut W, v: &Value) -> io::Result<()> {
    match v {
        Value::Null => w.write_all(&[0xc0]),
        Value::Bool(true) => w.write_all(&[0xc3]),
        Value::Bool(false) => w.write_all(&[0xc2]),
        Value::Number(n) => {
            // Canonical: non-negative -> smallest unsigned; negative -> smallest
            // signed; non-integers -> float64.
            if let Some(u) = n.as_u64() {
                write_uint(w, u)
            } else if let Some(i) = n.as_i64() {
                write_neg_int(w, i)
            } else {
                let f = n.as_f64().unwrap();
                w.write_all(&[0xcb])?;
                w.write_all(&f.to_be_bytes())
            }
        }
        Value::String(s) => write_str(w, s),
        Value::Array(arr) => {
            write_array_header(w, arr.len())?;
            for e in arr {
                write_value(w, e)?;
            }
            Ok(())
        }
        Value::Object(map) => {
            write_map_header(w, map.len())?;
            for (k, val) in map {
                write_str(w, k)?;
                write_value(w, val)?;
            }
            Ok(())
        }
    }
}

fn write_neg_int<W: Write>(w: &mut W, v: i64) -> io::Result<()> {
    // v is always negative here (non-negatives go through write_uint).
    if v >= -32 {
        w.write_all(&[(v & 0xff) as u8])
    } else if v >= -0x80 {
        w.write_all(&[0xd0, v as i8 as u8])
    } else if v >= -0x8000 {
        w.write_all(&[0xd1])?;
        w.write_all(&(v as i16).to_be_bytes())
    } else if v >= -0x8000_0000 {
        w.write_all(&[0xd2])?;
        w.write_all(&(v as i32).to_be_bytes())
    } else {
        w.write_all(&[0xd3])?;
        w.write_all(&v.to_be_bytes())
    }
}

fn write_uint<W: Write>(w: &mut W, v: u64) -> io::Result<()> {
    if v <= 0x7f {
        w.write_all(&[v as u8])
    } else if v <= 0xff {
        w.write_all(&[0xcc, v as u8])
    } else if v <= 0xffff {
        w.write_all(&[0xcd])?;
        w.write_all(&(v as u16).to_be_bytes())
    } else if v <= 0xffff_ffff {
        w.write_all(&[0xce])?;
        w.write_all(&(v as u32).to_be_bytes())
    } else {
        w.write_all(&[0xcf])?;
        w.write_all(&v.to_be_bytes())
    }
}

fn write_str<W: Write>(w: &mut W, s: &str) -> io::Result<()> {
    let body = s.as_bytes();
    let n = body.len();
    if n <= 31 {
        w.write_all(&[0xa0 | n as u8])?;
    } else if n <= 0xff {
        w.write_all(&[0xd9, n as u8])?;
    } else if n <= 0xffff {
        w.write_all(&[0xda])?;
        w.write_all(&(n as u16).to_be_bytes())?;
    } else {
        w.write_all(&[0xdb])?;
        w.write_all(&(n as u32).to_be_bytes())?;
    }
    w.write_all(body)
}

fn write_array_header<W: Write>(w: &mut W, n: usize) -> io::Result<()> {
    if n <= 15 {
        w.write_all(&[0x90 | n as u8])
    } else if n <= 0xffff {
        w.write_all(&[0xdc])?;
        w.write_all(&(n as u16).to_be_bytes())
    } else {
        w.write_all(&[0xdd])?;
        w.write_all(&(n as u32).to_be_bytes())
    }
}

fn write_map_header<W: Write>(w: &mut W, n: usize) -> io::Result<()> {
    if n <= 15 {
        w.write_all(&[0x80 | n as u8])
    } else if n <= 0xffff {
        w.write_all(&[0xde])?;
        w.write_all(&(n as u16).to_be_bytes())
    } else {
        w.write_all(&[0xdf])?;
        w.write_all(&(n as u32).to_be_bytes())
    }
}

/// Stream a collection as concatenated MessagePack values, one per element.
pub fn encode_array<'a, I, W>(items: I, writer: &mut W) -> io::Result<usize>
where
    I: IntoIterator<Item = &'a Value>,
    W: Write,
{
    let mut count = 0usize;
    for v in items {
        write_value(writer, v)?;
        count += 1;
    }
    Ok(count)
}

// --- decoding --------------------------------------------------------------

/// A lazy iterator over concatenated MessagePack values read from a reader.
pub struct MsgPackIter<R: Read> {
    reader: R,
    buf: Vec<u8>,
    pos: usize,
    eof: bool,
    finished: bool,
}

/// Stream concatenated MessagePack values, yielding one at a time.
pub fn decode_array<R: Read>(reader: R) -> MsgPackIter<R> {
    MsgPackIter {
        reader,
        buf: Vec::new(),
        pos: 0,
        eof: false,
        finished: false,
    }
}

impl<R: Read> MsgPackIter<R> {
    fn fill(&mut self, need: usize) -> io::Result<bool> {
        while self.buf.len() - self.pos < need {
            // Drop the consumed prefix so memory stays bounded.
            if self.pos > 0 {
                self.buf.drain(0..self.pos);
                self.pos = 0;
            }
            let mut tmp = [0u8; 65536];
            let k = self.reader.read(&mut tmp)?;
            if k == 0 {
                self.eof = true;
                return Ok(self.buf.len() - self.pos >= need);
            }
            self.buf.extend_from_slice(&tmp[..k]);
        }
        Ok(true)
    }

    fn at_end(&mut self) -> io::Result<bool> {
        if self.pos < self.buf.len() {
            return Ok(false);
        }
        if self.eof {
            return Ok(true);
        }
        Ok(!self.fill(1)?)
    }

    fn take(&mut self, n: usize) -> io::Result<Vec<u8>> {
        if !self.fill(n)? {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "flatwire msgpack: truncated value",
            ));
        }
        let out = self.buf[self.pos..self.pos + n].to_vec();
        self.pos += n;
        Ok(out)
    }

    fn u8(&mut self) -> io::Result<u8> {
        Ok(self.take(1)?[0])
    }

    fn be16(&mut self) -> io::Result<u16> {
        let b = self.take(2)?;
        Ok(u16::from_be_bytes([b[0], b[1]]))
    }
    fn be32(&mut self) -> io::Result<u32> {
        let b = self.take(4)?;
        Ok(u32::from_be_bytes([b[0], b[1], b[2], b[3]]))
    }
    fn be64(&mut self) -> io::Result<[u8; 8]> {
        let b = self.take(8)?;
        Ok([b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]])
    }

    fn read_value(&mut self) -> io::Result<Value> {
        let c = self.u8()?;
        if c <= 0x7f {
            return Ok(Value::from(c as i64));
        }
        if c >= 0xe0 {
            return Ok(Value::from((c as i8) as i64));
        }
        if (0x80..=0x8f).contains(&c) {
            return self.read_map((c & 0x0f) as usize);
        }
        if (0x90..=0x9f).contains(&c) {
            return self.read_array((c & 0x0f) as usize);
        }
        if (0xa0..=0xbf).contains(&c) {
            let s = self.take((c & 0x1f) as usize)?;
            return Ok(Value::String(str_utf8(s)?));
        }
        match c {
            0xc0 => Ok(Value::Null),
            0xc2 => Ok(Value::Bool(false)),
            0xc3 => Ok(Value::Bool(true)),
            0xca => {
                let b = self.take(4)?;
                let f = f32::from_be_bytes([b[0], b[1], b[2], b[3]]) as f64;
                Ok(num_f64(f))
            }
            0xcb => {
                let b = self.be64()?;
                Ok(num_f64(f64::from_be_bytes(b)))
            }
            0xcc => Ok(Value::from(self.u8()? as i64)),
            0xcd => Ok(Value::from(self.be16()? as i64)),
            0xce => Ok(Value::from(self.be32()? as i64)),
            0xcf => Ok(Value::from(u64::from_be_bytes(self.be64()?))),
            0xd0 => Ok(Value::from((self.u8()? as i8) as i64)),
            0xd1 => Ok(Value::from(self.be16()? as i16 as i64)),
            0xd2 => Ok(Value::from(self.be32()? as i32 as i64)),
            0xd3 => Ok(Value::from(i64::from_be_bytes(self.be64()?))),
            0xd9 => {
                let n = self.u8()? as usize;
                let s = self.take(n)?;
                Ok(Value::String(str_utf8(s)?))
            }
            0xda => {
                let n = self.be16()? as usize;
                let s = self.take(n)?;
                Ok(Value::String(str_utf8(s)?))
            }
            0xdb => {
                let n = self.be32()? as usize;
                let s = self.take(n)?;
                Ok(Value::String(str_utf8(s)?))
            }
            0xdc => {
                let n = self.be16()? as usize;
                self.read_array(n)
            }
            0xdd => {
                let n = self.be32()? as usize;
                self.read_array(n)
            }
            0xde => {
                let n = self.be16()? as usize;
                self.read_map(n)
            }
            0xdf => {
                let n = self.be32()? as usize;
                self.read_map(n)
            }
            // bin family -> represent as an array of byte numbers is lossy; the
            // JSON model has no bytes, so we surface bin as a string of raw bytes
            // decoded lossily is wrong. flatwire's JSON model never emits bin, so
            // treat bin as unsupported on decode to stay honest.
            0xc4 | 0xc5 | 0xc6 => Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "flatwire msgpack: binary (bin) type is not part of the JSON value model",
            )),
            _ => Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("flatwire msgpack: unknown prefix 0x{c:02x}"),
            )),
        }
    }

    fn read_array(&mut self, n: usize) -> io::Result<Value> {
        let mut arr = Vec::with_capacity(n);
        for _ in 0..n {
            arr.push(self.read_value()?);
        }
        Ok(Value::Array(arr))
    }

    fn read_map(&mut self, n: usize) -> io::Result<Value> {
        let mut map = Map::new();
        for _ in 0..n {
            let k = self.read_value()?;
            let key = match k {
                Value::String(s) => s,
                other => other.to_string(),
            };
            let v = self.read_value()?;
            map.insert(key, v);
        }
        Ok(Value::Object(map))
    }
}

fn str_utf8(b: Vec<u8>) -> io::Result<String> {
    String::from_utf8(b).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
}

fn num_f64(f: f64) -> Value {
    serde_json::Number::from_f64(f)
        .map(Value::Number)
        .unwrap_or(Value::Null)
}

impl<R: Read> Iterator for MsgPackIter<R> {
    type Item = io::Result<Value>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.finished {
            return None;
        }
        match self.at_end() {
            Ok(true) => {
                self.finished = true;
                None
            }
            Ok(false) => Some(self.read_value()),
            Err(e) => {
                self.finished = true;
                Some(Err(e))
            }
        }
    }
}
